Implement TorchIO transforms wrapper analogous to TorchVision transfo…

SomeUserName1 opened the pull request:
…rms wrapper and test case

Fixes #7499  .

### Description
As discussed in the issue, this PR implements a wrapper class for TorchIO transforms, analogous to the TorchVision transforms wrapper.

The test cases just check that transforms are callable and that after applying a transform, the result is different from the inputs.

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

SomeUserName1 commented:
Not sure where I have to add the torchio dependency to make CI pass.

KumoLiu commented:
Hi @SomeUserName1, you can add a skip wrapper like this:
https://github.com/Project-MONAI/MONAI/blob/c86e790d56f3a5ad07db6e642c6a65e7779b7b21/tests/test_patch_wsi_dataset.py#L184

Thanks.

SomeUserName1 commented:
One test failed, i have no idea why as the test and corresponding code is nothing that I've touched

```
Traceback (most recent call last):
  File "/home/runner/work/MONAI/MONAI/tests/test_regularization.py", line 73, in test_cutmixd
    self.assertTrue(not torch.allclose(output["a"], output["b"]))
AssertionError: False is not true
```

KumoLiu commented:
Looks like a random error, fixed in this commit.
https://github.com/Project-MONAI/MONAI/pull/7542/commits/40bb1300e8baccecbe05a61726ba508d27753eb2

SomeUserName1 commented:
@KumoLiu only the blossom-ci check hasn't passed. Can we merge?

KumoLiu commented:
Hi @SomeUserName1, thanks for the PR! We might consider incorporating this new feature after the 1.3.1 release due to the new requirements included in this PR.

johnzielke reviewed (commented):
I found this by accident, so first of all: Really nice, I'd love to be able to easily use some of the transforms in torchio, and I think it helps the available transforms there gain visibility.

While looking at this, I had a few thoughts:
1. It would be nice to have this available for the dictionary versions as well (like [torchvisiond](https://docs.monai.io/en/stable/transforms.html#torchvisiond))
2. The transform does currently not implement the RandomizableTrait, so I'm not sure how this would interact with CacheDatasets etc. I think it might help to have a version with the RandomizableTrait for all Random transforms (it seems like these are easy to find since they start with "Random") and one without for the rest. I'm not sure how to best design this, but one could maybe use the `__new__()` method to select the correct one based on the name. Alternatively an error/warning could be thrown if the wrong version is used.

SomeUserName1 commented on monai/transforms/utility/array.py:
```diff
@@ -1163,6 +1164,42 @@ def __call__(self, img: NdarrayOrTensor):
+    """
+    This is a wrapper transform for TorchIO transforms based on the specified transform name and args.
+    As most of the TorchIO transforms only work for PyTorch Tensor, this transform expects input
+    data to be PyTorch Tensor, users can easily call `ToTensor` transform to convert a Numpy array to Tensor.
```
The TorchVision class also does the conversion. I documented it for both.

johnzielke replied:
Will this then return the original type again? If it does not (e.g. always return torch.tensor), maybe add a comment about this in the docs.

SomeUserName1 replied:
From the tio docs of tio.Transform.__call__:

> Transform data and return a result of the same type.

SomeUserName1 commented on monai/transforms/utility/array.py:
```diff
@@ -1163,6 +1164,42 @@ def __call__(self, img: NdarrayOrTensor):
         return out
 
 
+class TorchIO:
```
Added Transform to both as base class

SomeUserName1 commented:
>     1. It would be nice to have this available for the dictionary versions as well (like [torchvisiond](https://docs.monai.io/en/stable/transforms.html#torchvisiond))

Implemented.

>     2. The transform does currently not implement the RandomizableTrait, so I'm not sure how this would interact with CacheDatasets etc. I think it might help to have a version with the RandomizableTrait for all Random transforms (it seems like these are easy to find since they start with "Random") and one without for the rest. I'm not sure how to best design this, but one could maybe use the `__new__()` method to select the correct one based on the name. Alternatively an error/warning could be thrown if the wrong version is used.

They also have a common RandomTransform base class in tio, which differs from MONAIs base RandomTransform.
All tio transforms are probabilistic and support the `p=` parameter specifying the probability that the transform is applied. This is analogous to RandomizableTransform in MONAI.
The RandomTransform in tio states that there is randomness in the applied transform itself (like RandomNoise), additionally that its application (or not application) is probabilistic.

Thus, I add the RandomizableTrait as ABC to the TorchIO wrapper for all transforms. 

I'm not sure if I should add Randomizable or RandomTranform as well as this is implemented in torchio already. So adding these two would be clearer in terms of the capabilities of these transforms but require manipulating TorchIO objects/class instances in MONAI code introducing complexity with little practical value (probabilistic application can be done via the tio parameters; is there a benefit in adding the same functionality via MONAI with the different concepts in mind?)

SomeUserName1 commented:
Random I think, didnt touch regularization.
```
 ======================================================================
FAIL: test_cutmixd (tests.test_regularization.TestCutMix)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/MONAI/MONAI/tests/test_regularization.py", line 73, in test_cutmixd
    self.assertTrue(not torch.allclose(output["a"], output["b"]))
AssertionError: False is not true

----------------------------------------------------------------------
```

SomeUserName1 commented on monai/transforms/utility/array.py:
```diff
@@ -1163,6 +1162,36 @@ def __call__(self, img: NdarrayOrTensor):
+        """
+        Args:
+            name: The transform name in TorchIO package.
+            args: parameters for the TorchIO transform.
```
Is adding a ref to the TorchIO docs acceptable too? 
Would ease maintainability, if tio changes sth.

johnzielke replied:
I would do both, but for such a basic parameter, I would at least mention it's name here.

SomeUserName1 commented on monai/transforms/utility/dictionary.py:
```diff
@@ -1419,6 +1423,36 @@ def __call__(self, data: Mapping[Hashable, NdarrayOrTensor]) -> dict[Hashable, N
+    def __call__(self, data: Mapping[Hashable, NdarrayOrTensor]) -> dict[Hashable, NdarrayOrTensor]:
+        d = dict(data)
+        for key in self.key_iterator(d):
+            d[key] = self.trans(d[key])
```
So the TorchIO transform is instantiated once, i.e. the same object is used for all `d[key]`. 
As the randomness is handled per Transform, it is the same for all applications of the transform.

johnzielke replied:
But doesnt't the torchvision transform apply e.g. a different random flip each time it's called?

SomeUserName1 replied:
So far I've only used it with subjects, for which the above statement is true.
In our most common use-case, you're absolutely right, as we're passing `torch.tensor`s and not `tio.subject`s by default.

tio handles that via the Subjects entity and by using the RandomTransform just to generate params that are then passed to a non-random version (e.g. `RandomAffine` samples params and then consturcts `Affine` ). The latter is actually applied to all images of a subject to achieve consistency.

If multiple `torch.tensor`s are passed, you're right, it's a different one for each `torch.tensor`. TorchIO is only able to apply the same transform if a `tio.Subject` is passed.

TorchVision samples the parameters directly in the forward call, i.e. applies the transform with the sampled parameters and does not store them. 

So I think the behavior is the same for both torchvision and torchio, that when using random transforms, different transform params are sampled and thus a transform with different params is applied.

To achieve consistent transforms per dict, I could add a flag, which then wraps the images in the dict into a `tio.Subject`.

Is this reasonable?

johnzielke replied:
If torchio provides this system, it would be nice to be able to use the subject or any other option to achieve this for the dictionary.
That sounds like a great option! (The 4Ds are channel+3spatial dims i'm assuming?) I'd make this flag true by default, since this is the behavior that is common in Monai Transforms.

I don't think this PR needs to fix that for torchvision, but you could add something about it in the docs (and maybe a warning if you want), so it's easy to spot in the future.

KumoLiu commented on monai/transforms/utility/array.py:
```diff
@@ -1163,6 +1162,47 @@ def __call__(self, img: NdarrayOrTensor):
+                 or dict containing 4D tensors as values
+
+        """
+        return self.trans(img)
```
If we want to include this transform in the MONAI transform chain, I guess we should ensure the output type which can be compatible with the existing transforms. Perhaps we can add a flag here, what do you think?

ericspod replied:
For the randomised transforms the deterministic interface that we use in MONAI should be used here as well if possible. This includes setting the random state with `randomize` (like [here](https://github.com/Project-MONAI/MONAI/blob/4029c422f1da84cdf5e4ce546dbff33245cbe025/monai/transforms/spatial/array.py#L1475)) and using `set_random_state` to set the state. This may be difficult with how TorchIO transforms work but we can do as much as feasible here.

SomeUserName1 replied:
Okay so for all inherently random use RandomizableTrait and for the inherently deterministic ones use the deterministic interface.
Can you provide me a link to the deterministic interface/an example that uses it, please?

ericspod replied:
[RandRotate90](https://github.com/Project-MONAI/MONAI/blob/4029c422f1da84cdf5e4ce546dbff33245cbe025/monai/transforms/spatial/array.py#L1191) may be a good example of how to do things in a simple format. We use `randomize` to compute random state the transform relies on, then apply it in `__call__` as needed. [RandRotate90d](https://github.com/Project-MONAI/MONAI/blob/4029c422f1da84cdf5e4ce546dbff33245cbe025/monai/transforms/spatial/dictionary.py#L672) demonstrates doing this for dictionary based transforms.

KumoLiu commented on monai/transforms/utility/dictionary.py:
```diff
@@ -1419,6 +1423,53 @@ def __call__(self, data: Mapping[Hashable, NdarrayOrTensor]) -> dict[Hashable, N
+        self,
+        keys: KeysCollection,
+        name: str,
+        apply_same_transform: bool = False,
```
The same comments about the RandomizableTrait inherit with the array version.
And what's this `apply_same_transform` arg used for? In MONAI, we always apply the same transform on the  sequence keys, such as image and label. Is there a typical case here users may want apply different transforms on the different keys? If not, perhaps we can remove this one.

SomeUserName1 replied:
Apply same is indeed a bit misleading.

I had a use case with random affine, where the same random affine should be applied to both input and output for augmentation with patches.

Same/similar for random elastic deformations. 

Can remove if you want to, would constrain the applicable augmentations in a sense or restrict the net to learn to output only a certain orientation/affine, i.e. all inputs would need registration to the data with which the net was trained.

SomeUserName1 replied:
It's different in TorchIO so I had to hack arround it. The `apply_same = false` is merely to reproduce TorchIOs behavior.

KumoLiu reviewed (commented):
Sorry for the delayed review; the point release is finally done. I have some concerns about the support for randomness, leave the comments inline.

@ericspod could you please also help review the PR? Thanks!

ericspod commented on monai/transforms/utility/array.py:
```diff
@@ -1163,6 +1162,47 @@ def __call__(self, img: NdarrayOrTensor):
         return out
 
 
+class TorchIO(Transform, RandomizableTrait):
```
I think we do need this yes, otherwise we're going to explicitly state that only the randomised TorchIO transforms are going to be used here. This probably isn't ideal even though most of the deterministic transforms in TorchIO are represented in native MONAI transforms. Let's have a `TorchIO` and `RandTorchIO` pair of transforms and separate out which are deterministic or not.

SomeUserName1 replied:
How would the transform behave with caching if p!=1 in the tio transform [source of tio.Transform](https://github.com/fepegar/torchio/blob/13407ab7f6d562a0953fa6642ce198f297129589/src/torchio/transforms/transform.py#L97C2-L97C22) ?

I get the distinction between probabilistic in terms of application probability (like cropping with p!=1) and randomness inherent in the transform (like random noise) and agree with adding RandTorchIO for the latter case, but am unsure how to deal with the former.

ericspod replied:
We do have a fundamental conflict here in that our transforms differentiate between random ones and deterministic ones, in both type and interface, whereas the interface for TorchIO does not. We may have to rely on the user only using non-random TorchIO transforms with this `TorchIO` class and any randomised ones use `RandTorchIO`. The alternative is that we keep `RandomizableTrait` here and then TorchIO preprocessing transforms essentially can't be used in a deterministic way, ie. can't be cached by `CacheDataset` amongst other things. If all of the deterministic behaviour of TorchIO is covered by our existing transforms I'd say this is fine, but if that's not fine then my first solution is maybe all we have and we'll have to trust the user to follow documentation stating how to use our wrappers.

SomeUserName1 replied:
Maybe you guys can discuss how to get the traits/interfaces compliant between tio and monai and then tell me how you'd like me to implement/adjust it.

My personal opinion is the one I implemented, i.e. caching makes no sense with probabilistic application or - put differently - mark the transform as random if randomness is involved no matter where the randomness comes from.

In case you want to introduce the distinction between randomness inside the transformation and probabilistic transform application into MonAI's interfaces/traits I'm also happy to help with that.

KumoLiu replied:
You might find this PR helpful for reference: [PR #5569](https://github.com/Project-MONAI/MONAI/pull/5569). There’s a discussion there that addresses application probability and inherent probability in more detail: [relevant discussion link](https://github.com/Project-MONAI/MONAI/pull/5569#discussion_r1030822630).

SomeUserName1 commented:
Is this something that you'd like to proceed with any time soon? otherwise I'd close the PR

KumoLiu commented:
Hi @SomeUserName1, apologies for overlooking this PR. I’ll review it and aim to have it merged by the end of the week. Thanks.

SomeUserName1 commented:
No worries!

I agree that the randomness issue still stands and will implement the latter version.
>  [...] or state in the docstring to use only randomised TorchIO transforms.

SomeUserName1 commented:
So, I added the changes as discussed.
Additionally, I added RandTorchVision as well to actually unify the wrappers as discussed in PR #5569 , mentioned by @KumoLiu.


Right now all RandFooD transforms apply a cached version of the randomized transformed. So e.g. RandomCrop will apply the same cropping to all elements in the dataset. Not sure how helpful this is for augmentation.
Instead in the RandFooD classes, the transforms should be called via RandFoo instead of Foo to again avoid caching.
However I'm not sure if I just don't get the use case.

ericspod commented on tests/test_torchiod.py:
```diff
@@ -0,0 +1,55 @@
+    @parameterized.expand(TEST2)
+    def test_random_transform(self, input_param, input_data):
+        set_determinism(seed=0)
+        result = TorchIOd(**input_param)(input_data)
```
This should use `RandTorchIOd` if we're going to continue to have both random and non-random TorchIO transforms. What I had stated earlier was essentially to have only `RandTorchIO/RandTorchIOd` classes and only use randomised TorchIO transforms, so not use `RescaleIntensity`  as in the previous test which is failing for some reason. If you really want to keep `TorchIO/TorchIOd` in addition to `RandTorchIO/RandTorchIOd` that's fine but you should use the correct class here.

SomeUserName1 replied:
Sure, will do. 
Also sorry for the mess I created yesterday wrt. the pipelines, it was rather late.

SomeUserName1 commented:
Locally the 4 unit test classes i wrote ran through, mypy and the formatters passed, so I hope this will be the last time, I'll kick off the pipeline. 
If the pipeline runs green, please review.

Thanks for your time @ericspod & @KumoLiu !

ericspod commented on docs/source/transforms.rst:
The other failing test is related to an actual bug in the test itself which will go away if the doc issue is sorted.

SomeUserName1 commented:
isort adds the indentation that sphinx complains about...

ericspod reviewed (approved):
I think this is good now, thanks!

KumoLiu commented:
/build
